enhancements/20230316-multi-csi-support.md

```diff
@@ -0,0 +1,141 @@
+
+2. Allow the user to install additional packages during installation through the Harvester config file.
+
+Currently, Harvester OS does not support users adding additional software packages.
+
+[`elemental-toolkit`](https://rancher.github.io/elemental-toolkit/docs/) provides a stage called `after-install-chroot`. Commands executed in this stage do not encounter file system write issues, and the results of user commands are persistent and will not be lost due to a reboot. We can use this mechanism and expose it through Harvester Configuration.
+
+A config example for install rpm package in Harvester:
+
+```yaml
+os:
+  afterInstallChrootCommands:
+    - rpm -ivh <the url of rpm package>
+```
+
+In the `after-install-chroot` stage, dns resolution is not available.
+
+If users need to access a domain name, users need to create a temporary `/etc/resolv.conf` file at first, for example:
+
+```yaml
+os:
+  afterInstallChrootCommands:
+    - "rm -f /etc/resolv.conf && echo 'nameserver 8.8.8.8' | sudo tee /etc/resolv.conf"
+    - "mkdir /usr/local/bin"
+    - "curl -fsSL -o get_helm.sh https://raw.githubusercontent.com/helm/helm/main/scripts/get-helm-3 && chmod 700 get_helm.sh && ./get_helm.sh"
+    - "rm -f /etc/resolv.conf && ln -s /var/run/netconfig/resolv.conf /etc/resolv.conf"
+```
+
+#### Harvester UI
+1. Changes to **StorageClass** page.
+- When creating a StorageClass, support selecting other provisioners. UI can get options from K8s's `csidrivers` API. 
+- If it is a non-Longhorn provisioner, only show the general parameters.
+![fake-storageclass-edit.png](20230316-multi-csi-support/fake-storageclass-edit.png)
+![fake-storageclass-detail.png](20230316-multi-csi-support/fake-storageclass-detail.png)
+
+2. Setting `csi-driver-config` Details Page
+- The default content of setting `csi-driver-config`:
+```json
+{
+  "driver.longhorn.io": {
+    "volumeSnapshotClassName": "longhorn-snapshot",
+    "backupVolumeSnapshotClassName": "longhorn"
+  }
+}
+```
+- The first level map key is the name of the CSI driver. UI can also get options from K8s's `csidrivers` API.
+
+![csi-driver-config-overview.png](20230316-multi-csi-support/csi-driver-config-overview.png)
+![csi-driver-config-edit.png](20230316-multi-csi-support/csi-driver-config-edit.png)
+
+### Test plan
+
+https://github.com/harvester/harvester/issues/2405#issuecomment-1549634520
+https://github.com/harvester/harvester/issues/3689#issuecomment-1563130124
+
+### Upgrade strategy
+N/A
+
+## Note
+- Installing OS packages for other CSI drivers will depend on https://github.com/rancher/elemental-toolkit/issues/1739.
+- Changes to the OS in the `after-install-chroot` stage will be lost after the Harvester upgrade, so user also need to configure the `after-upgrade-chroot` stage before upgrading Harvester.
```

Comment:
Do we have a GH issue tracking this one?

Reply:
HEP updated.
We do not need to expose the `after-upgrade-chroot` through Harvester config. User only need to configure the `after-upgrade-chroot` stage by adding an custom oem configuration file to `/oem` dir refer to https://rancher.github.io/elemental-toolkit/docs/customizing/runtime_persistent_changes/ before upgrading.